File: sdks/php/src/Api/Suppressions.php
<?php

declare(strict_types=1);

namespace Apollo\Deploy\Signal\Sdk\Api;

use Apollo\Deploy\Signal\Sdk\SDKError;
use Apollo\Deploy\Signal\Sdk\SSEEvent;
use Apollo\Deploy\Signal\Sdk\Transport;
use Apollo\Deploy\Signal\Sdk\AddSuppressionBody;
use Apollo\Deploy\Signal\Sdk\SuppressionImportBody;
use Apollo\Deploy\Signal\Sdk\SuppressionImportResponse;
use Apollo\Deploy\Signal\Sdk\SuppressionPageResponse;
use Apollo\Deploy\Signal\Sdk\SuppressionResponse;
use Psr\Http\Message\ResponseInterface;

/** SuppressionsAPI domain operations. */
class Suppressions
{
    public function __construct(private Transport $transport)
    {
    }

    /**
     * @param string $projectId
     * @return SuppressionPageResponse
     * @throws SDKError
     */
    public function listSuppressions(
        string $projectId,
    ): SuppressionPageResponse
    {
        $path = '/v1/projects/${projectId}/suppressions';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(SuppressionPageResponse::class, 'fromArray')) {
            return SuppressionPageResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @return void
     * @throws SDKError
     */
    public function exportSuppressions(
        string $projectId,
    ): void
    {
        $path = '/v1/projects/${projectId}/suppressions/export';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'GET',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $this->transport->readResponseBody($response);
        return;
    }

    /**
     * @param string $projectId
     * @param AddSuppressionBody $input
     * @return SuppressionResponse
     * @throws SDKError
     */
    public function addSuppression(
        string $projectId,
        AddSuppressionBody $input,
    ): SuppressionResponse
    {
        $path = '/v1/projects/${projectId}/suppressions';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(SuppressionResponse::class, 'fromArray')) {
            return SuppressionResponse::fromArray($data);
        }
        return $data;
    }

    /**
     * @param string $projectId
     * @param string $email
     * @return void
     * @throws SDKError
     */
    public function removeSuppression(
        string $projectId,
        string $email,
    ): void
    {
        $path = '/v1/projects/${projectId}/suppressions/${email}';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);
        $path = str_replace('${email}', rawurlencode(Transport::wireValue($email)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'DELETE',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];

        $response = $this->transport->executeRequest($request);
        $this->transport->readResponseBody($response);
        return;
    }

    /**
     * @param string $projectId
     * @param SuppressionImportBody $input
     * @return SuppressionImportResponse
     * @throws SDKError
     */
    public function importSuppressions(
        string $projectId,
        SuppressionImportBody $input,
    ): SuppressionImportResponse
    {
        $path = '/v1/projects/${projectId}/suppressions/import';
        $path = str_replace('${projectId}', rawurlencode(Transport::wireValue($projectId)), $path);

        $query = [];

        $headers = [];

        $request = [
            'method' => 'POST',
            'path' => $path,
            'query' => $query,
            'headers' => $headers,
        ];
        if ($input !== null) {
            $request['body'] = $input;
            $request['contentType'] = 'application/json';
        }

        $response = $this->transport->executeRequest($request);
        $data = $this->decodeResponse($response);
        if (is_array($data) && method_exists(SuppressionImportResponse::class, 'fromArray')) {
            return SuppressionImportResponse::fromArray($data);
        }
        return $data;
    }

    private function decodeResponse(ResponseInterface $response): mixed
    {
        $contents = $this->transport->readResponseBody($response);
        if (trim($contents) === '') {
            $requestId = $response->getHeaderLine('x-request-id')
                ?: $response->getHeaderLine('request-id')
                ?: null;
            throw new SDKError(
                message: 'Expected a response body for a non-void operation',
                status: $response->getStatusCode(),
                code: 'RESPONSE_BODY_MISSING',
                requestId: $requestId,
            );
        }

        try {
            return json_decode($contents, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            return $contents;
        }
    }
}

File: sdks/rust/src/api/suppressions.rs
use crate::error::SdkError;
use crate::transport::Transport;
use reqwest::Method;
use serde::Serialize;
use crate::types::{ AddSuppressionBody, SuppressionImportBody, SuppressionImportResponse, SuppressionPageResponse, SuppressionResponse };

/// SuppressionsAPI API operations.
pub struct SuppressionsAPI {
    transport: Transport,
}

impl SuppressionsAPI {
    pub(crate) fn new(transport: Transport) -> Self {
        Self { transport }
    }

    pub async fn list_suppressions(
        &self,
        project_id: String,
    ) -> Result<SuppressionPageResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/suppressions".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<SuppressionPageResponse>(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await
    }

    pub async fn export_suppressions(
        &self,
        project_id: String,
    ) -> Result<(), SdkError> {
        let mut path = "/v1/projects/${projectId}/suppressions/export".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_raw(
            Method::GET,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await.map(|_| ())
    }

    pub async fn add_suppression(
        &self,
        project_id: String,
        body: &AddSuppressionBody,
    ) -> Result<SuppressionResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/suppressions".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<SuppressionResponse>(
            Method::POST,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

    pub async fn remove_suppression(
        &self,
        project_id: String,
        email: String,
    ) -> Result<(), SdkError> {
        let mut path = "/v1/projects/${projectId}/suppressions/${email}".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));
        path = path.replace("${email}", &encode_path(&email.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute_raw(
            Method::DELETE,
            &path,
            query,
            None::<&serde_json::Value>,
            headers,
            None,
        ).await.map(|_| ())
    }

    pub async fn import_suppressions(
        &self,
        project_id: String,
        body: &SuppressionImportBody,
    ) -> Result<SuppressionImportResponse, SdkError> {
        let mut path = "/v1/projects/${projectId}/suppressions/import".to_string();
        path = path.replace("${projectId}", &encode_path(&project_id.to_string()));

        let mut query_params: Vec<(String, String)> = Vec::new();
        let query = (!query_params.is_empty()).then_some(query_params.as_slice());

        let mut headers: Vec<(String, String)> = Vec::new();
        let headers = (!headers.is_empty()).then_some(headers.as_slice());

        self.transport.execute::<SuppressionImportResponse>(
            Method::POST,
            &path,
            query,
            Some(body),
            headers,
            None,
        ).await
    }

}

fn encode_path(value: &str) -> String {
    urlencoding::encode(value).into_owned()
}

fn encode_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    match serde_json::to_value(value)
        .map_err(|error| SdkError::ValidationError(error.to_string()))?
    {
        serde_json::Value::String(value) => Ok(value),
        value => Ok(value.to_string()),
    }
}

fn encode_cookie_name(name: &str) -> String {
    urlencoding::encode(name).into_owned()
}

fn encode_cookie_value<T: Serialize>(value: &T) -> Result<String, SdkError> {
    let value = encode_value(value)?;
    Ok(urlencoding::encode(&value).into_owned())
}
